src/crush/CrushCompiler.cc

```diff
@@ -668,6 +671,8 @@ int CrushCompiler::parse_bucket(iter_t const& i)
 
 int CrushCompiler::parse_rule(iter_t const& i)
 {
+  crush.populate_classes();
```

Comment:
maybe it's better to instead put this in parse_crush, guarded by a bool did_first_rule; or similar, so it's only done once before the first rule (and at the end if there were no rules).